From 3a6406a76b785f8ee521eeec34172874d23f2407 Mon Sep 17 00:00:00 2001
From: Chunguang Xu <brookxu@tencent.com>
Date: Tue, 6 Jul 2021 05:47:26 +0800
Subject: [PATCH 4581/4585] [Backport] block: fix the problem of io_ticks
 becoming smaller

mainline inclusion
from mainline-v5.14-rc1
commit d80c228d44640f0b47b57a2ca4afa26ef87e16b0
category: bugfix
bugzilla: 187475

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d80c228d44640f0b47b57a2ca4afa26ef87e16b0

--------------------------------

On the IO submission path, blk_account_io_start() may interrupt
the system interruption. When the interruption returns, the value
of part->stamp may have been updated by other cores, so the time
value collected before the interruption may be less than part->
stamp. So when this happens, we should do nothing to make io_ticks
more accurate? For kernels less than 5.0, this may cause io_ticks
to become smaller, which in turn may cause abnormal ioutil values.

Signed-off-by: Chunguang Xu <brookxu@tencent.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Link: https://lore.kernel.org/r/1625521646-1069-1-git-send-email-brookxu.cn@gmail.com
Signed-off-by: Jens Axboe <axboe@kernel.dk>

conflict:
	block/blk-core.c

Signed-off-by: Li Nan <linan122@huawei.com>
---
 block/blk-core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/blk-core.c b/block/blk-core.c
index 7d4324f6e664..a18cfc467d41 100644
--- a/block/blk-core.c
+++ b/block/blk-core.c
@@ -1276,7 +1276,7 @@ void update_io_ticks(struct hd_struct *part, unsigned long now, bool end)
 	unsigned long stamp;
 again:
 	stamp = READ_ONCE(part->stamp);
-	if (unlikely(stamp != now) &&
+	if (unlikely(time_after(now, stamp)) &&
 		likely(cmpxchg(&part->stamp, stamp, now) == stamp)) {
 		if (precise_iostat) {
 			if (end || part_in_flight(part))
-- 
2.17.1

